File: lua/pay_kit/protocols/x402/exact/verify.lua
--[[
x402 SVM-exact 11-rule structural verifier (Lua port).

Mirrors the Ruby reference at
`ruby/lib/x402/protocol/schemes/exact/verify.rb` and the Rust spine
at `rust/crates/x402/src/protocol/schemes/exact/verify.rs`. Raises
the same canonical reject strings the cross-language harness
substring-matches against.

Rules:
  1. Instruction count 3..=6                    (spine verify.rs:230-235)
  2. ix[0] = ComputeBudget SetComputeUnitLimit  (verify.rs:240-248)
  3. ix[1] = ComputeBudget SetComputeUnitPrice <= MAX (verify.rs:250-264)
  4. ix[2] = SPL TransferChecked                (verify.rs:380-410)
  5. Authority guard (no fee-payer in transfer auth) (verify.rs:382)
  6. Mint match                                 (verify.rs:395-400)
  7. Destination ATA match (re-derive)          (verify.rs:402-405)
  8. Amount match                               (verify.rs:407-410)
  9. ix[3..6] in allowlist (Memo + Lighthouse ONLY; ATA-create rejected)
 10. Memo binding (exactly one if extra.memo set)
 11. Token program strict bind to extra.tokenProgram

Reuses pay_kit.solana (PayCore) for transaction parsing, ATA derive,
and base58. Ed25519 client-signature verification routes through
pay_kit.util.ed25519 so the openssl / luasodium backend choice
is consistent with the rest of the SDK.
]]

local base64 = require('pay_kit.util.base64_std')
local base58 = require('pay_kit.solana.base58')
local tx_mod = require('pay_kit.solana.transaction')
local ata    = require('pay_kit.solana.ata')
local ed25519 = require('pay_kit.util.ed25519')
local uint    = require('pay_kit.util.uint')

local M = {}

local COMPUTE_BUDGET_PROGRAM    = 'ComputeBudget111111111111111111111111111111'
local MEMO_PROGRAM              = 'MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr'
-- Official x402 SVM exact Lighthouse program id (specs/schemes/exact/
-- scheme_exact_svm.md), matching the PHP (Verifier::LIGHTHOUSE_PROGRAM)
-- and Go (lighthouseProgram) verifiers. The prior `L1TEVtgA75k...` value
-- was wrong and would have rejected wallet-injected Lighthouse guards.
local LIGHTHOUSE_PROGRAM        = 'L2TExMFKdjpN9kozasaurPirfHy9P8sbXoAN1qA3S95'
local TOKEN_PROGRAM             = 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA'
local TOKEN_2022_PROGRAM        = 'TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb'
-- Mirrors the Rust spine constant
-- (rust/crates/x402/src/protocol/schemes/exact/verify.rs:17). The prior
-- 50_000 value rejected canonical wallet transactions whose compute-unit
-- price legitimately sits above 50k but under the protocol cap.
local MAX_COMPUTE_UNIT_PRICE_MICROLAMPORTS = 5000000

-- Multiply an unsigned-decimal string by a small integer (< 2^31).
local function mul_small(decimal, factor)
  local carry, out = 0, {}
  for i = #decimal, 1, -1 do
    local product = tonumber(decimal:sub(i, i)) * factor + carry
    out[#out + 1] = tostring(product % 10)
    carry = math.floor(product / 10)
  end
  while carry > 0 do
    out[#out + 1] = tostring(carry % 10)
    carry = math.floor(carry / 10)
  end
  local chars = {}
  for idx = #out, 1, -1 do chars[#chars + 1] = out[idx] end
  local text = table.concat(chars):gsub('^0+', '')
  return text == '' and '0' or text
end

-- Read a little-endian u64 from a binary string at offset 1-based and
-- return it as an exact decimal string. Mirrors the Rust spine's
-- `u64::from_le_bytes` (verify.rs:405-409 / :350-354): a float-based
-- reconstruction loses precision above 2^53, so a malicious amount or
-- compute-unit price in the high u64 range would round to a different
-- value than the one signed on-chain. Decode byte-wise so the full u64
-- range is exact.
local function read_u64_le(s, start)
  if not s or #s < start + 7 then
    error('invalid_exact_svm_payload_no_transfer_instruction')
  end
  -- Accumulate big-endian: total = total * 256 + byte, from MSB to LSB.
  local total = '0'
  for offset = 7, 0, -1 do
    total = uint.add(mul_small(total, 256), tostring(s:byte(start + offset)))
  end
  return total
end

-- Look up the program id (base58 string) for an instruction.
local function program_of(account_keys, ix)
  local idx = ix.program_id_index + 1
  return account_keys[idx]
end

local function account_at(account_keys, ix, slot)
  return account_keys[ix.accounts[slot + 1] + 1]
end

-- Rule 2: ix[0] = ComputeBudget SetComputeUnitLimit.
local function verify_compute_limit(ix, account_keys)
  local data = ix.data
  if program_of(account_keys, ix) ~= COMPUTE_BUDGET_PROGRAM
      or #data ~= 5 or data:byte(1) ~= 2 then
    error('invalid_exact_svm_payload_transaction_instructions_compute_limit_instruction')
  end
end

-- Rule 3: ix[1] = ComputeBudget SetComputeUnitPrice <= MAX.
local function verify_compute_price(ix, account_keys)
  local data = ix.data
  if program_of(account_keys, ix) ~= COMPUTE_BUDGET_PROGRAM
      or #data ~= 9 or data:byte(1) ~= 3 then
    error('invalid_exact_svm_payload_transaction_instructions_compute_price_instruction')
  end
  local micro = read_u64_le(data, 2)
  if uint.compare(micro, tostring(MAX_COMPUTE_UNIT_PRICE_MICROLAMPORTS)) > 0 then
    error('invalid_exact_svm_payload_transaction_instructions_compute_price_instruction_too_high')
  end
end

-- Decode a base58 requirement field (e.g. asset, payTo) once.
local function b58_field(requirement, key)
  local v = requirement[key]
  if type(v) ~= 'string' or v == '' then
    error('invalid_exact_svm_payload_missing_field_' .. key)
  end
  return v
end

local function string_extra(requirement, key, required)
  local extra = requirement.extra or {}
  local v = extra[key]
  if (v == nil or v == '') and required then
    error('invalid_exact_svm_payload_missing_extra_' .. key)
  end
  return v
end

-- Rules 4 + 5 + 6 + 7 + 8 + 11 in one pass on the TransferChecked
-- instruction.
local function verify_transfer(ix, account_keys, requirement, managed_signers)
  local program = program_of(account_keys, ix)
  -- Bind the transfer program to the canonical SPL token program set,
  -- NOT to `extra.tokenProgram`. Mirrors the Rust spine
  -- (verify.rs:373): the program id is accepted iff it is TOKEN_PROGRAM
  -- or TOKEN_2022_PROGRAM, derived from the actual instruction. A
  -- canonical offer may omit `extra.tokenProgram`, so requiring it would
  -- reject a spec-valid credential the Rust verifier accepts.
  if program ~= TOKEN_PROGRAM and program ~= TOKEN_2022_PROGRAM then
    error('invalid_exact_svm_payload_no_transfer_instruction')
  end
  local data = ix.data
  if #ix.accounts < 4 or #data ~= 10 or data:byte(1) ~= 12 then
    error('invalid_exact_svm_payload_no_transfer_instruction')
  end
  local source      = account_at(account_keys, ix, 0)
  local mint        = account_at(account_keys, ix, 1)
  local destination = account_at(account_keys, ix, 2)
  local authority   = account_at(account_keys, ix, 3)

  -- Rule 5: authority guard.
  for i = 1, #managed_signers do
    if managed_signers[i] == authority or managed_signers[i] == source then
      error('invalid_exact_svm_payload_transaction_fee_payer_transferring_funds')
    end
  end
  for j = 1, #ix.accounts do
    local key = account_keys[ix.accounts[j] + 1]
    for i = 1, #managed_signers do
      if managed_signers[i] == key then
        error('invalid_exact_svm_payload_transaction_fee_payer_in_instruction_accounts')
      end
    end
  end

  -- Rule 6: mint match.
  local expected_mint = b58_field(requirement, 'asset')
  if mint ~= expected_mint then
    error('invalid_exact_svm_payload_mint_mismatch')
  end

  -- Rule 7: destination ATA match.
  local pay_to = b58_field(requirement, 'payTo')
  local expected_destination = ata.derive(pay_to, expected_mint, program)
  if destination ~= expected_destination then
    error('invalid_exact_svm_payload_recipient_mismatch')
  end

  -- Rule 8: amount match. Compare as exact unsigned decimals so a
  -- full-range u64 amount cannot collide with a different value through
  -- float rounding (mirrors Rust's u64 equality, verify.rs:414).
  local amount = read_u64_le(data, 2)
  local expected_amount = requirement.amount
  if expected_amount == nil or expected_amount == '' then
    expected_amount = requirement.maxAmountRequired
  end
  if type(expected_amount) ~= 'string' or expected_amount == '' then
    error('invalid_exact_svm_payload_amount_mismatch')
  end
  if uint.compare(amount, expected_amount) ~= 0 then
    error('invalid_exact_svm_payload_amount_mismatch')
  end

  return {
    program     = program,
    source      = source,
    mint        = mint,
    destination = destination,
    authority   = authority,
    amount      = amount,
  }
end

local function find_memo_match(account_keys, instructions, expected_memo)
  local memo_count, last_memo_data = 0, nil
  for i = 4, #instructions do
    if program_of(account_keys, instructions[i]) == MEMO_PROGRAM then
      memo_count = memo_count + 1
      last_memo_data = instructions[i].data
    end
  end
  if memo_count ~= 1 then
    error('invalid_exact_svm_payload_memo_count')
  end
  if last_memo_data ~= expected_memo then
    error('invalid_exact_svm_payload_memo_mismatch')
  end
end

-- Top-level: verify a base64-encoded transaction against a server
-- offer. `managed_signers` is the list of base58 keys the server has
-- authority over (the operator's pubkey at minimum); the verifier
-- refuses to treat any of those as the transfer source / authority
-- so a malicious credential cannot "spend the facilitator's funds".
function M.verify(transaction_b64, requirement, managed_signers)
  local decode_ok, raw = pcall(base64.decode, transaction_b64)
  if not decode_ok or not raw or raw == '' then
    error('invalid_exact_svm_payload_base64')
  end
  local ok, parsed_or_err = pcall(tx_mod.from_bytes, raw)
  if not ok then error('invalid_exact_svm_payload_transaction_parse') end
  local parsed = parsed_or_err

  local instructions = parsed.message.instructions
  if #instructions < 3 or #instructions > 6 then
    error('invalid_exact_svm_payload_transaction_instructions_length')
  end

  verify_compute_limit(instructions[1], parsed.message.account_keys)
  verify_compute_price(instructions[2], parsed.message.account_keys)
  local transfer = verify_transfer(instructions[3], parsed.message.account_keys,
                                   requirement, managed_signers)

  -- Rule 9: ix[3..6] allowlist. Optional slots may carry ONLY Lighthouse
  -- (wallet-injected guard) or SPL Memo. An Associated-Token-Program
  -- ATA-create is NOT permitted: per the official x402 SVM exact contract
  -- the destination ATA MUST pre-exist. Lighthouse is allowed in ANY
  -- optional slot because wallets inject a variable number of guards
  -- (Phantom 1, Solflare 2). Mirrors php Verifier and go verify.go.
  local reasons = {
    'invalid_exact_svm_payload_unknown_fourth_instruction',
    'invalid_exact_svm_payload_unknown_fifth_instruction',
    'invalid_exact_svm_payload_unknown_sixth_instruction',
  }
  for i = 4, #instructions do
    local ix = instructions[i]
    local program = program_of(parsed.message.account_keys, ix)
    local slot_index = i - 4  -- 0-based offset within slots 3..5
    local allowed = (program == MEMO_PROGRAM) or (program == LIGHTHOUSE_PROGRAM)
    if not allowed then
      error(reasons[slot_index + 1] or 'invalid_exact_svm_payload_unknown_optional_instruction')
    end
  end

  -- Rule 10: memo binding.
  local expected_memo = string_extra(requirement, 'memo', false)
  if expected_memo and expected_memo ~= '' then
    find_memo_match(parsed.message.account_keys, instructions, expected_memo)
  end

  return transfer
end

-- Verify non-managed client signatures on the transaction envelope.
-- Mirrors Ruby `verify_client_signatures!`. The transaction bytes
-- carry [signatures... | message]; the server-managed signers
-- (facilitator pubkey) are skipped because the facilitator signs
-- after verification.
function M.verify_client_signatures(transaction_b64, managed_signer_b58_list)
  local decode_ok, raw = pcall(base64.decode, transaction_b64)
  if not decode_ok or not raw or raw == '' then
    error('invalid_exact_svm_payload_signature')
  end

  -- compact-u16 short-vec at offset 0 = signature count.
  local function read_short_vec(bytes, offset)
    local value, shift = 0, 0
    local i = offset + 1
    repeat
      local b = bytes:byte(i)
      if not b then error('invalid_exact_svm_payload_signature') end
      value = value + (b % 128) * (2 ^ shift)
      i = i + 1
      shift = shift + 7
    until b < 128
    return value, i - 1
  end

  local signature_count, signatures_offset = read_short_vec(raw, 0)
  local message_offset = signatures_offset + (signature_count * 64)
  if message_offset >= #raw then
    error('invalid_exact_svm_payload_signature')
  end
  local message = raw:sub(message_offset + 1)
  if message:byte(1) ~= 0x80 then
    error('invalid_exact_svm_payload_signature')
  end
  local required_signatures = message:byte(2)
  if required_signatures > signature_count then
    error('invalid_exact_svm_payload_signature')
  end
  local account_count, account_offset = read_short_vec(message, 4)
  if required_signatures > account_count then
    error('invalid_exact_svm_payload_signature')
  end

  local zero_sig = string.rep(string.char(0), 64)
  local managed_set = {}
  for i = 1, #managed_signer_b58_list do
    managed_set[managed_signer_b58_list[i]] = true
  end

  for index = 0, required_signatures - 1 do
    local signer_key_start = account_offset + (index * 32) + 1
    if signer_key_start + 31 > #message then
      error('invalid_exact_svm_payload_signature')
    end
    local signer_key_bytes = message:sub(signer_key_start, signer_key_start + 31)
    local signer_key_b58 = base58.encode(signer_key_bytes)
    if not managed_set[signer_key_b58] then
      local sig = raw:sub(signatures_offset + (index * 64) + 1,
                          signatures_offset + (index * 64) + 64)
      if sig == zero_sig then
        error('invalid_exact_svm_payload_signature')
      end
      local ok = ed25519.verify(signer_key_bytes, message, sig)
      if not ok then
        error('invalid_exact_svm_payload_signature')
      end
    end
  end
end

return M

File: python/src/pay_kit/protocols/x402/exact/verify.py
"""x402 ``exact`` self-hosted 11-rule structural verifier and constants.

``ExactVerifier`` follows the Rust spine rule-for-rule and reject-code-for-
reject-code (``rust/crates/x402/src/protocol/schemes/exact/verify.rs`` and the
server backstops at ``rust/crates/x402/src/server/exact.rs``), adding only
strictly-stronger defensive rejects; cross-checked against the PHP port at
``php/src/Protocols/X402/Exact/Verifier.php``.
"""

from __future__ import annotations

import base64
import struct
from typing import Any, cast

from pay_kit._paycore.mints import derive_ata
from pay_kit.errors import InvalidProofError

__all__ = [
    "ExactVerifier",
    "EXACT_SCHEME",
    "X402_VERSION",
    "X402_VERSION_V1",
    "X402_VERSION_V2",
    "COMPUTE_BUDGET_PROGRAM",
    "MEMO_PROGRAM",
    "LIGHTHOUSE_PROGRAM",
    "TOKEN_2022_PROGRAM",
    "MAX_COMPUTE_UNIT_PRICE",
]

#: x402 ``exact`` scheme identifier (rust ``EXACT_SCHEME``, types.rs:6).
EXACT_SCHEME = "exact"

#: Legacy x402 protocol version. Mirrors the rust spine constant
#: ``X402_VERSION_V1`` (rust/crates/x402/src/constants.rs:10).
X402_VERSION_V1 = 1
#: Canonical x402 protocol version used by current payments. Mirrors the rust
#: spine ``X402_VERSION_V2`` (constants.rs:13).
X402_VERSION_V2 = 2
#: Default x402 protocol version emitted in challenges and required on the
#: canonical credential path. Aliases the canonical (v2) version so existing
#: call sites keep emitting/expecting the default producer's wire.
X402_VERSION = X402_VERSION_V2

#: ComputeBudget program id (instruction[0]/[1] guard).
COMPUTE_BUDGET_PROGRAM = "ComputeBudget111111111111111111111111111111"
#: SPL Memo program id (allowlisted optional instruction + memo binding).
MEMO_PROGRAM = "MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr"
#: Lighthouse assertion program id (allowlisted optional instruction).
#: Must match the rust spine constant ``LIGHTHOUSE_PROGRAM`` in
#: ``rust/crates/x402/src/protocol/schemes/exact/types.rs``.
LIGHTHOUSE_PROGRAM = "L2TExMFKdjpN9kozasaurPirfHy9P8sbXoAN1qA3S95"
#: Token-2022 program id (accepted transfer program alongside the route's).
TOKEN_2022_PROGRAM = "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb"
#: Maximum SetComputeUnitPrice in microlamports. Matches the Rust spine
#: constant ``MAX_COMPUTE_UNIT_PRICE_MICROLAMPORTS`` in verify.rs.
MAX_COMPUTE_UNIT_PRICE = 5_000_000


def _u64_le(data: bytes, offset: int) -> int:
    """Read a little-endian u64 at ``offset``; reject on a short buffer."""
    if len(data) < offset + 8:
        raise InvalidProofError(
            "invalid_exact_svm_payload_no_transfer_instruction",
            code="invalid_exact_svm_payload_no_transfer_instruction",
        )
    return struct.unpack_from("<Q", data, offset)[0]


class ExactVerifier:
    """Structural 11-rule verifier for the x402 SVM ``exact`` scheme.

    Decodes a standard-base64 (padded) versioned transaction and confirms it
    matches the route's advertised requirement. Each failure raises
    :class:`InvalidProofError` carrying the canonical reject string the
    cross-language harness substring-matches against. Mirrors the
    rust ``verify_exact_instructions`` ordering exactly.
    """

    @staticmethod
    def verify(
        transaction_base64: str,
        requirement: dict[str, Any],
        managed_signers: list[str],
    ) -> dict[str, Any]:
        """Verify a base64 transaction against the route's x402 requirement.

        ``requirement`` is one ``accepts[]`` entry (the server offer).
        ``managed_signers`` lists server-managed pubkeys (typically the
        facilitator fee payer) that must never be the transfer authority.
        Returns a dict describing the matched transfer on success.
        """
        from solders.transaction import VersionedTransaction

        try:
            raw = base64.b64decode(transaction_base64, validate=True)
        except Exception as exc:  # noqa: BLE001 - any decode failure is a reject
            raise InvalidProofError(
                "invalid_exact_svm_payload_base64",
                code="invalid_exact_svm_payload_base64",
            ) from exc
        if not raw:
            raise InvalidProofError(
                "invalid_exact_svm_payload_base64",
                code="invalid_exact_svm_payload_base64",
            )

        try:
            tx = VersionedTransaction.from_bytes(raw)
        except Exception as exc:  # noqa: BLE001
            raise InvalidProofError(
                "invalid_exact_svm_payload_transaction_parse",
                code="invalid_exact_svm_payload_transaction_parse",
            ) from exc

        message = tx.message
        instructions = list(message.instructions)
        account_keys = [str(key) for key in message.account_keys]

        # Rule 1: instruction count 3..=6.
        n = len(instructions)
        if n < 3 or n > 6:
            raise InvalidProofError(
                "invalid_exact_svm_payload_transaction_instructions_length",
                code="invalid_exact_svm_payload_transaction_instructions_length",
            )

        # Rule 2: ix[0] = ComputeBudget SetComputeUnitLimit (disc 2, 5 bytes).
        ExactVerifier._verify_compute_limit(instructions[0], account_keys)
        # Rule 3: ix[1] = ComputeBudget SetComputeUnitPrice (disc 3, 9 bytes, <= MAX).
        ExactVerifier._verify_compute_price(instructions[1], account_keys)
        # Rules 4 + 5 + 6 + 7 + 8 + 11: transferChecked.
        transfer = ExactVerifier._verify_transfer(instructions[2], account_keys, requirement, managed_signers)

        # Rule 9: ix[3:] allowlist. Per the official x402 SVM exact contract
        # (specs/schemes/exact/scheme_exact_svm.md), the only permitted optional
        # instructions are Lighthouse (wallet-injected user-protection asserts,
        # Phantom=1 / Solflare=2) and SPL Memo. A Create-ATA / Associated Token
        # Program instruction is NOT allowed: the destination ATA MUST pre-exist
        # (Rule 7 derives and pins the destination ATA). This matches the
        # Rust/Go verifiers, which accept Lighthouse or Memo in ANY optional
        # slot (rust verify.rs iter().skip(3); go verify.go case Memo/Lighthouse
        # for all i>=3) and never accept ATA-create in this shape. Lighthouse is
        # not slot-restricted because wallets inject a variable number of guards.
        reasons = (
            "invalid_exact_svm_payload_unknown_fourth_instruction",
            "invalid_exact_svm_payload_unknown_fifth_instruction",
            "invalid_exact_svm_payload_unknown_sixth_instruction",
        )
        for i in range(3, n):
            ix = instructions[i]
            program = ExactVerifier._program_of(account_keys, ix)
            slot_index = i - 3
            allowed = program in (MEMO_PROGRAM, LIGHTHOUSE_PROGRAM)
            if not allowed:
                reason = (
                    reasons[slot_index]
                    if slot_index < len(reasons)
                    else "invalid_exact_svm_payload_unknown_optional_instruction"
                )
                raise InvalidProofError(reason, code=reason)

        # Rule 10: memo binding (exactly one Memo == extra.memo if set).
        expected_memo = ExactVerifier._string_extra(requirement, "memo", required=False)
        if expected_memo:
            ExactVerifier._find_memo_match(account_keys, instructions, expected_memo)

        # The destination ATA must pre-exist; ATA-create is never accepted.
        transfer["destinationCreateAta"] = False
        return transfer

    @staticmethod
    def _verify_compute_limit(ix: Any, account_keys: list[str]) -> None:
        program = ExactVerifier._program_of(account_keys, ix)
        data = bytes(ix.data)
        if program != COMPUTE_BUDGET_PROGRAM or len(data) != 5 or data[0] != 2:
            raise InvalidProofError(
                "invalid_exact_svm_payload_transaction_instructions_compute_limit_instruction",
                code="invalid_exact_svm_payload_transaction_instructions_compute_limit_instruction",
            )

    @staticmethod
    def _verify_compute_price(ix: Any, account_keys: list[str]) -> None:
        program = ExactVerifier._program_of(account_keys, ix)
        data = bytes(ix.data)
        if program != COMPUTE_BUDGET_PROGRAM or len(data) != 9 or data[0] != 3:
            raise InvalidProofError(
                "invalid_exact_svm_payload_transaction_instructions_compute_price_instruction",
                code="invalid_exact_svm_payload_transaction_instructions_compute_price_instruction",
            )
        micro = _u64_le(data, 1)
        if micro > MAX_COMPUTE_UNIT_PRICE:
            reason = "invalid_exact_svm_payload_transaction_instructions_compute_price_instruction_too_high"
            raise InvalidProofError(reason, code=reason)

    @staticmethod
    def _verify_transfer(
        ix: Any,
        account_keys: list[str],
        requirement: dict[str, Any],
        managed_signers: list[str],
    ) -> dict[str, Any]:
        program = ExactVerifier._program_of(account_keys, ix)
        # Rule 11: token program strict bind to extra.tokenProgram.
        token_program_extra = ExactVerifier._string_extra(requirement, "tokenProgram", required=True)
        if program != token_program_extra and program != TOKEN_2022_PROGRAM:
            raise InvalidProofError(
                "invalid_exact_svm_payload_no_transfer_instruction",
                code="invalid_exact_svm_payload_no_transfer_instruction",
            )
        data = bytes(ix.data)
        # solders CompiledInstruction.accounts is a list of u8 account indices;
        # solders ships no stubs, so annotate the shape explicitly at the boundary.
        accounts: list[int] = [int(a) for a in ix.accounts]
        # Rule 4: transferChecked shape (disc 12, 10-byte data, >= 4 accounts).
        if len(accounts) < 4 or len(data) != 10 or data[0] != 12:
            raise InvalidProofError(
                "invalid_exact_svm_payload_no_transfer_instruction",
                code="invalid_exact_svm_payload_no_transfer_instruction",
            )

        source = ExactVerifier._account_at(account_keys, ix, 0)
        mint = ExactVerifier._account_at(account_keys, ix, 1)
        destination = ExactVerifier._account_at(account_keys, ix, 2)
        authority = ExactVerifier._account_at(account_keys, ix, 3)

        # Rule 5: authority guard (no managed signer as authority/source/account).
        for managed in managed_signers:
            if managed in (authority, source):
                raise InvalidProofError(
                    "invalid_exact_svm_payload_transaction_fee_payer_transferring_funds",
                    code="invalid_exact_svm_payload_transaction_fee_payer_transferring_funds",
                )
        for idx in accounts:
            key = account_keys[idx] if 0 <= idx < len(account_keys) else None
            if key is None:
                continue
            for managed in managed_signers:
                if managed == key:
                    raise InvalidProofError(
                        "invalid_exact_svm_payload_transaction_fee_payer_in_instruction_accounts",
                        code="invalid_exact_svm_payload_transaction_fee_payer_in_instruction_accounts",
                    )

        # Rule 6: mint match (offer carries the resolved on-chain mint on `asset`).
        expected_mint = ExactVerifier._b58_field(requirement, "asset")
        if mint != expected_mint:
            raise InvalidProofError(
                "invalid_exact_svm_payload_mint_mismatch",
                code="invalid_exact_svm_payload_mint_mismatch",
            )

        # Rule 7: destination ATA match (re-derive owner+mint+token_program).
        pay_to = ExactVerifier._b58_field(requirement, "payTo")
        expected_destination = derive_ata(pay_to, expected_mint, program)
        if destination != expected_destination:
            raise InvalidProofError(
                "invalid_exact_svm_payload_recipient_mismatch",
                code="invalid_exact_svm_payload_recipient_mismatch",
            )

        # Rule 8: amount match (u64 LE at data[1:9]).
        amount = _u64_le(data, 1)
        expected_amount = ExactVerifier._amount_field(requirement)
        if amount != expected_amount:
            raise InvalidProofError(
                "invalid_exact_svm_payload_amount_mismatch",
                code="invalid_exact_svm_payload_amount_mismatch",
            )

        return {
            "program": program,
            "source": source,
            "mint": mint,
            "destination": destination,
            "authority": authority,
            "amount": amount,
        }

    @staticmethod
    def _find_memo_match(account_keys: list[str], instructions: list[Any], expected_memo: str) -> None:
        count = 0
        last_data: bytes | None = None
        for i in range(3, len(instructions)):
            ix = instructions[i]
            if ExactVerifier._program_of(account_keys, ix) == MEMO_PROGRAM:
                count += 1
                last_data = bytes(ix.data)
        if count != 1:
            raise InvalidProofError(
                "invalid_exact_svm_payload_memo_count",
                code="invalid_exact_svm_payload_memo_count",
            )
        if last_data is None or last_data.decode("utf-8", "replace") != expected_memo:
            raise InvalidProofError(
                "invalid_exact_svm_payload_memo_mismatch",
                code="invalid_exact_svm_payload_memo_mismatch",
            )

    @staticmethod
    def _program_of(account_keys: list[str], ix: Any) -> str:
        idx = int(ix.program_id_index)
        return account_keys[idx] if 0 <= idx < len(account_keys) else ""

    @staticmethod
    def _account_at(account_keys: list[str], ix: Any, slot: int) -> str:
        accounts = list(ix.accounts)
        if slot >= len(accounts):
            raise InvalidProofError(
                "invalid_exact_svm_payload_no_transfer_instruction",
                code="invalid_exact_svm_payload_no_transfer_instruction",
            )
        idx = int(accounts[slot])
        return account_keys[idx] if 0 <= idx < len(account_keys) else ""

    @staticmethod
    def _b58_field(requirement: dict[str, Any], key: str) -> str:
        value = requirement.get(key)
        if not isinstance(value, str) or value == "":
            raise InvalidProofError(
                f"invalid_exact_svm_payload_missing_field_{key}",
                code=f"invalid_exact_svm_payload_missing_field_{key}",
            )
        return value

    @staticmethod
    def _string_extra(requirement: dict[str, Any], key: str, *, required: bool) -> str | None:
        extra = requirement.get("extra")
        value = cast("dict[str, object]", extra).get(key) if isinstance(extra, dict) else None
        if (value is None or value == "") and required:
            raise InvalidProofError(
                f"invalid_exact_svm_payload_missing_extra_{key}",
                code=f"invalid_exact_svm_payload_missing_extra_{key}",
            )
        return value if isinstance(value, str) else None

    @staticmethod
    def _amount_field(requirement: dict[str, Any]) -> int:
        value = requirement.get("amount")
        if value is None:
            value = requirement.get("maxAmountRequired")
        if not isinstance(value, (str, int)):
            raise InvalidProofError(
                "invalid_exact_svm_payload_missing_field_amount",
                code="invalid_exact_svm_payload_missing_field_amount",
            )
        return int(value)
